packages/cerebral-router/src/router.js

```diff
@@ -0,0 +1,148 @@
+
+    controller.on('initialized', () => {
+      this.routesConfig = flattenConfig(options.routes)
+      this.routableSignals = getRoutableSignals(this.routesConfig, controller)
+
+      addressbar.on('change', this.onUrlChange.bind(this))
+      controller.runTree.on('start', this.onSignalStart.bind(this))
+      controller.on('flush', this.onFlush.bind(this))
+  
+      if (!options.preventAutostart) {
+        this.onUrlChange()
+      }
+    })
+  }
+
+  onUrlChange(event) {
+    let url = event ? event.target.value : this.addressbar.value
+    url = url.replace(this.addressbar.origin, '')
+
+    if (this.options.onlyHash && !~url.indexOf('#')) {
+      // treat hash absense as root route
+      url = url + '#/'
+    }
+
+    // check if url should be routed
+    if (url.indexOf(this.options.baseUrl) === 0) {
+      const {match, route, values} = this.mapper.map(url.replace(this.options.baseUrl, ''), this.routesConfig) || {}
+
+      if (match) {
+        console.log(match)
+        const {signal, map} = match
+
+        event && event.preventDefault()
+        this.addressbar.value = url
+
+        // TODO: move to flattenConfig
+        const mapByTarget = Object.keys(map || {}).reduce((mapped, key) => {
+          const {target, path} = map[key]
+
+          if (['input', 'state'].indexOf(target) === -1) {
+            // TODO: improve errror message
+            throw new Error('Cerebral router - unknown target provided to map option')
+          }
+
+          mapped[target][key] = path
+
+          return mapped
+        }, {input:{}, state:{}})
+
+        if (Object.keys(mapByTarget.state).length) {
+          this.controller.runTree('router.routed', [
+            ({state}) => {
+              Object.keys(mapByTarget.state).forEach((key) => state.set(mapByTarget.state[key], values[key]))
+            }
+          ])
+          this.activeStateMap = Object.assign(match, {mapByTarget, route})
+        } else {
+          this.activeStateMap = null
+        }
+
+        if (signal) {
+          /*
+            TODO: transform payload according map to input if provided
+          */
+          this.controller.getSignal(signal)(values)
+        }
+      } else {
+        if (this.options.allowEscape) return
+
+        event && event.preventDefault()
+        console.warn(`Cerebral router - No route matched ${url}, navigation was prevented. Please verify url or catch unmatched routes with a "/*" route.`) // eslint-disable-line no-console
+      }
+    }
+  }
+
+  onSignalStart(execution, payload) {
+    const signal = this.routableSignals[execution.name]
+    if (signal) {
+      const route = signal.route
+      const input = payload
+
+      this.addressbar.value = this.options.baseUrl + this.mapper.stringify(route, input)
+    }
+  }
+
+  onFlush(changed) {
+    if (!this.activeStateMap) {
+      return
+    }
+    const {mapByTarget, route} = this.activeStateMap
+    if (
+      Object.keys(mapByTarget.state).find((key) => {
```

Comment:
@Guria No utility, but without gotten a chance to review code I would suggest converting every state dependency path to an object of:

```js
{
  foo: true,
  bar: {
    baz: true
  }
}
```

And when you get the `change` object, which is similar, you traverse it and see if you hit something. 

But yeah, it is only used once inside dependency store which both components and computed uses, so did not create a utility for it